_posts/2018-05-02-load-balanced-smart-proxies-with-haproxy.md

```diff
@@ -0,0 +1,151 @@
+
+### Generating Puppet certificates
+On the first smart proxy, generate the Puppet certificates by running
+
+```shell
+puppet cert generate proxy02.example.net --dns_alt_names=smartproxy.example.net
+```
+
+Install the `puppetserver` package on the second smart proxy, to create the `puppet` user and the directory structure in `/etc/puppetlabs`.
+
+Now copy `/etc/puppetlabs/puppet/ssl/*/proxy02.example.net.pem` and `/etc/puppetlabs/puppet/ssl/certs/ca.pem` to the second smart proxy and ensure ownership and permissions are identical to those the first host. Also don't forget to check the SELinux contexts of the files you copied over (hint: `ls -lZ`) :)
+
+Repeat this step for all smart proxies that are going to sit behind the same load balancer.
+
+### Preparing `custom-hiera.yaml`
+These steps are identical to the steps used on the first Smart Proxy.
+
+### Running the installer
+As only the first Smart Proxy will host a PuppetCA, we point the Puppet agent to it explicitly during the installation and also configure the Puppet certificate to include the `smartproxy.example.net` name:
+```
+[root@smartproxy02] # foreman-installer \
+…
+--puppet-dns-alt-names smartproxy.example.net \
+--puppet-ca-server proxy01.example.net \
+--foreman-proxy-puppetca false \
+--puppet-server-ca false
+```
+## Installing HAProxy
+
+Please note that if you are using a different load balancer technology, you can use this configuration as reference.
+
+Configure SELinux to run in `permissive` mode.
+
+The HAProxy setup itself is super simple, just load-balance a bunch of ports in TCP mode round-robin across all the Smart Proxies. TCP mode is required because the load balancer will not be doing TLS termination. There are just two exceptions to that: port 443 should use `source` not `roundrobin` to not confuse Yum, and the additional port for the PuppetCA (8141) is forwarded only to the first smart proxy.
+
+| Service    | Port | Mode | Balance mode | Destination |
+| ---------- | ---- | ---- | ------------ | ----------- |
+| HTTP       | 80   | TCP  | `roundrobin` | port 80 on all proxies |
+| Anaconda   | 8000 | TCP  | `roundrobin` | port 8000 on all proxies|
+| HTTPS      | 443  | TCP  | `source`     | port 443 on all proxies |
+| RHSM       | 8443 | TCP  | `roundrobin` | port 8443 on all proxies |
+| AMQP       | 5647 | TCP  | `roundrobin` | port 5647 on all proxies |
+| Puppet     | 8140 | TCP  | `roundrobin` | port 8140 on all proxies |
+| PuppetCA   | 8141 | TCP  | `roundrobin` | port 8140 on the first proxy |
+| SmartProxy | 9090 | TCP  | `roundrobin` | port 9090 on all proxies |
+| Docker     | 5000 | TCP  | `roundrobin` | port 5000 on all proxies |
+
+[You can find an example `haproxy.cfg` in the `forklift` repository.](https://github.com/theforeman/forklift/blob/master/roles/haproxy/templates/haproxy.cfg.j2)
```

Comment:
I wonder if linking to a persistent location (i.e. the file in a specific commit) is better (so the link won't die) or worse (won't receive updates).